cpp/src/reductions/reductions.cpp

```diff
@@ -176,7 +181,24 @@ std::unique_ptr<scalar> reduce(column_view const& col,
                                rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::reduce(col, agg, output_dtype, cudf::default_stream_value, mr);
+  return detail::reduce(col, agg, output_dtype, std::nullopt, cudf::default_stream_value, mr);
 }
 
+std::unique_ptr<scalar> reduce(column_view const& col,
+                               std::unique_ptr<reduce_aggregation> const& agg,
+                               data_type output_dtype,
+                               scalar const& init,
+                               rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(col.type() == init.type(), "column and initial value must be the same type");
+  if (!(agg->kind == aggregation::SUM || agg->kind == aggregation::PRODUCT ||
+        agg->kind == aggregation::MIN || agg->kind == aggregation::MAX ||
+        agg->kind == aggregation::ANY || agg->kind == aggregation::ALL)) {
+    CUDF_FAIL(
+      "Initial value is only supported for SUM, PRODUCT, MIN, MAX, ANY, and ALL aggregation types");
+  }
+
+  CUDF_FUNC_RANGE();
+  return detail::reduce(col, agg, output_dtype, &init, cudf::default_stream_value, mr);
```

Comment:
This PR is a great time to fix a related problem in the `reduce` code paths: the `CUDF_EXPECTS` checks should all be implemented in the `detail::` function. Two reasons:
1. We want to make these error checks a part of the `CUDF_FUNC_RANGE` so that profiling accurately reflects the cost of checking these expectations.
2. From what I can tell, most `detail::` code in libcudf implements `CUDF_EXPECTS` and other validation logic, rather than only checking that in the public API. These are not expensive checks and shouldn't hurt performance, but they do help with enforcing internal consistency.

Just move the `CUDF_EXPECTS` / `CUDF_FAIL` logic to `detail::reduce` (and similarly for segmented reductions) so that the body of this function is only `CUDF_FUNC_RANGE();` and `return detail::reduce(...);`.